pkg/asset/installconfig/installconfig.go

```diff
@@ -157,6 +157,14 @@ func (a *InstallConfig) finish(filename string) error {
 	}
 	if a.Config.Azure != nil {
 		a.Azure = icazure.NewMetadata(a.Config.Azure.CloudName, a.Config.Azure.ARMEndpoint)
+		defaultMachinePlatform := a.Config.Platform.Azure.DefaultMachinePlatform
+		if defaultMachinePlatform != nil && defaultMachinePlatform.OSDisk.DiskEncryptionSet != nil && defaultMachinePlatform.OSDisk.DiskEncryptionSet.SubscriptionID == "" {
+			session, err := a.Azure.Session()
+			if err != nil {
+				return errors.Wrap(err, "failed to fetch session")
+			}
+			a.Config.Platform.Azure.DefaultMachinePlatform.OSDisk.DiskEncryptionSet.SubscriptionID = session.Credentials.SubscriptionID
+		}
```

Comment:
This code block looks good to me, but this doesn't look like the right place to set this value: adding this code here makes it less readable and it would only get worse if we build on this pattern. 

If this is the best place in the code to have this value set, then I think we just need one layer of abstraction on top of this (such as a general function for setting defaults that other platforms could use as well) . 

I couldn't find any existing code that sets defaults in this manner, so I think it would be good to build it out. If that's too much for this bug fix, we could either set the field to required or put this in place temporarily.

Reply:
+1 to the idea of filling the install-config.yaml with defaults as its own step in the install process. Once filled we can use the install-config for validation and for creating the resources knowing that both operations are seeing exactly the same config.